volume recycler: Don't start a new recycler pod if one already exists.

jsafrane opened the pull request:
Recycling is a long duration process and when the recycler controller is restarted in the meantime, it should not start a new recycler pod if there is one already running.

This means that the recycler pod must have deterministic name based on name of the recycled PV, we then get name conflicts when creating the pod.

Two things need to be changed:
- recycler controller and recycler plugins must pass the PV.Name to place, where the pod is created. This is most of the patch and it should be pretty straightforward.
- create recycler pod with deterministic name and check "already exists" error.

When at it, remove useless 'resourceVersion' argument and make log messages starting with lowercase.

There is an unit test to check the behavior + there is an e2e test that checks that regular recycling is not broken (it does not try to run two recycler pods in parallel as the recycler is single-threaded now).

jsafrane commented:
@kubernetes/sig-storage

jsafrane commented:
Reabsed

thockin commented:
I ACK receipt of this I will get to is ASAP

thockin commented:
Do we want to merge this given how close your mega PR is to completion?

jsafrane commented:
@thockin, this PR is orthogonal to big persistent volume controller refactoring, it can be merged at any time. Actually, I would prefer to merge it sooner than later.

childsb commented:
+1 to merge before controller refactor.  This looks good to me, waiting on @thockin for final review.

jsafrane committed:
volume recycler: Don't start a new recycler pod if one already exists.

Recycling is a long duration process and when the recycler controller is
restarted in the meantime, it should not start a new recycler pod if there is
one already running.

This means that the recycler pod must have deterministic name based on name
of the recycled PV, we then get name conflicts when creating the pod.

Two things need to be changed:
- recycler controller and recycler plugins must pass the PV.Name to place,
  where the pod is created.

- create recycler pod with deterministic name and check "already exists" error.

When at it, remove useless 'resourceVersion' argument and make log messages
starting with lowercase.

jsafrane commented:
Rebased, @thockin, PTAL.

jsafrane commented:
@k8s-bot test this issue: #23342
e2e exploded spectacularly and this is the closest flake I was able to find:
https://console.cloud.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/25263/kubernetes-pull-build-test-e2e-gce/40820/

```
May 19 07:42:39.393: INFO: Condition Ready of node e2e-gce-master-3-minion-czcc is false instead of true. Reason: NodeStatusUnknown, message: Kubelet stopped posting node status.
```

thockin commented:
@k8s-bot test this: github issue #IGNORE

thockin commented:
@k8s-bot test this: github issue #IGNORE

thockin commented:
`FAIL	k8s.io/kubernetes/test/integration	407.971s`

@ixdy - how can I figure out what part of integration tests failed?  It's a hopeless mess of log output

thockin commented:
@k8s-bot test this: github issue #IGNORE

I have no idea which flake it is - the logs are utterly useless

thockin commented:
@k8s-bot test this: github issue #IGNORE

I have no idea which flake it is - the logs are utterly useless
